src/compiler/scala/tools/nsc/typechecker/Infer.scala

```diff
@@ -1146,11 +1132,17 @@ trait Infer extends Checkable {
     // (see #3692, where the type param T's bounds were set to > : T <: T, so that parts looped)
     // the changes are rolled back by restoreTypeBounds, but might be unintentionally observed in the mean time
     def instantiateTypeVar(tvar: TypeVar): Unit = {
-      val tparam                    = tvar.origin.typeSymbol
-      val TypeBounds(lo0, hi0)      = tparam.info.bounds
-      val tb @ TypeBounds(lo1, hi1) = instBounds(tvar)
-      val enclCase                  = context.enclosingCaseDef
-      def enclCase_s                = enclCase.toString.replaceAll("\\n", " ").take(60)
+      val tparam = tvar.origin.typeSymbol
+      val tparams = cloneSymbols(tvar.typeParams)
+      val targs = tparams.map(_.tpeHK)
+      val instType = toOrigin(genPolyType(tparams, appliedType(tvar.constr.inst, targs)))
+      val instDefined = isFullyDefined(instType)
+      def instBounds(bounds: List[Type]) = if (instDefined) List(instType) else bounds.map(toOrigin)
+      val TypeBounds(lo0, hi0) = tparam.info.bounds
```

Comment:
This should avoid some extra allocation: 
```suggestion
      val lo0 = tparam.info.lowerBound
      val hi0 = tparam.info.upperBound     
```

Reply:
That's not really an allocation. It's a pattern match (which should be optimized away by the compiler). That being said I think there is an opportunity to simplify the code. But I think something along the lines of:
```scala
val bounds = tparam.info.bounds
```
And then inline `bounds.lo` and `bounds.hi` since they are referenced only once anyway.

Reply:
> And then inline bounds.lo and bounds.hi since they are referenced only once anyway.

I spoke too soon, that's not true.

Reply:
> This should avoid some extra allocation:

You mean that the `.bounds` call allocates internally. I didn't know that. Will make the change as suggested.